File: LeetCode - Algorithm/Dynamic Programming/120.py
# 120. Triangle
# https://leetcode.com/problems/triangle/

from typing import List
import math


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        memo = [[triangle[0][0]]]

        for i in range(1, len(triangle)):
            memo.append([])

            for j in range(len(triangle[i])):
                memoValue = triangle[i][j]
                upperLayer = math.inf
                if j - 1 >= 0:
                    upperLayer = min(upperLayer, memo[i - 1][j - 1])

                if j < len(memo[i - 1]):
                    upperLayer = min(upperLayer, memo[i - 1][j])

                memo[i].append(memoValue + upperLayer)

        return min(memo[-1])

File: LeetCode - Algorithm/Dynamic Programming/120.js
// 120. Triangle
// https://leetcode.com/problems/triangle/

/**
 * @param {number[][]} triangle
 * @return {number}
 */
var minimumTotal = function (triangle) {
  const memo = [];
  for (let i = 0; i < triangle.length; i++) {
    memo.push([]);

    for (let j = 0; j < triangle[i].length; j++) {
      memo[i].push(triangle[i][j]);
    }
  }

  for (let i = 1; i < triangle.length; i++) {
    for (let j = 0; j < triangle[i].length; j++) {
      const upperLeft = memo[i - 1][j - 1] ?? Number.POSITIVE_INFINITY;
      const upperRight = memo[i - 1][j] ?? Number.POSITIVE_INFINITY;

      memo[i][j] = Math.min(upperLeft, upperRight) + memo[i][j];
    }
  }

  let result = Number.POSITIVE_INFINITY;
  for (let num of memo[memo.length - 1]) {
    result = Math.min(result, num);
  }

  return result;
};
